game-core/src/main/java/games/strategy/triplea/delegate/battle/casualty/AaCasualtySelector.java

```diff
@@ -114,265 +112,209 @@ private static CasualtyDetails getLowLuckAaCasualties(
   }
 
-  /**
-   * http://triplea.sourceforge.net/mywiki/Forum#nabble-td4658925%7Ca4658925 returns two lists, the
-   * first list is the air units that can be evenly divided into groups of 3 or 6 (depending on
-   * radar) the second list is all the air units that do not fit in the first list
-   */
-  private static Tuple<List<List<Unit>>, List<Unit>> categorizeLowLuckAirUnits(
-      final Collection<Unit> units, final int groupSize) {
-    final Collection<UnitCategory> categorizedAir =
-        UnitSeparator.categorize(units, null, false, true);
-    final List<List<Unit>> groupsOfSize = new ArrayList<>();
-    final List<Unit> toRoll = new ArrayList<>();
-    for (final UnitCategory uc : categorizedAir) {
-      final int remainder = uc.getUnits().size() % groupSize;
-      final int splitPosition = uc.getUnits().size() - remainder;
-      final List<Unit> group = new ArrayList<>(uc.getUnits().subList(0, splitPosition));
-      if (!group.isEmpty()) {
-        for (int i = 0; i < splitPosition; i += groupSize) {
-          final List<Unit> miniGroup = new ArrayList<>(uc.getUnits().subList(i, i + groupSize));
-          if (!miniGroup.isEmpty()) {
-            groupsOfSize.add(miniGroup);
-          }
-        }
-      }
-      toRoll.addAll(uc.getUnits().subList(splitPosition, uc.getUnits().size()));
+  private static CasualtyDetails calculateRolledAaCasualties(
+      final List<Unit> availableTargets,
+      final AaPowerStrengthAndRolls unitPowerAndRollsMap,
+      final DiceRoll dice,
+      final IDelegateBridge bridge) {
+
+    final int hits = dice.getHits();
+    final Collection<Unit> hitTargets;
+    if (unitPowerAndRollsMap.calculateTotalRolls() == availableTargets.size()
+        && hits < availableTargets.size()) {
+      // there is a roll for every target but not enough hits to kill all of the targets
+      // so no need to get a random set of units since all units will either have a hit
+      // or miss roll
+      hitTargets = findRolledTargets(availableTargets, dice);
+    } else if (hits < availableTargets.size()) {
+      // there isn't a roll for every target so need to randomly pick the target for each hit
+      hitTargets = findRandomTargets(availableTargets, bridge, hits);
+    } else {
+      // all targets were hit so add them all
+      hitTargets = availableTargets;
     }
-    return Tuple.of(groupsOfSize, toRoll);
+
+    return buildCasualtyDetails(availableTargets, hitTargets);
   }
 
-  private static int getTotalHitpointsLeft(final Unit unit) {
-    if (unit == null) {
-      return 0;
+  /** Find the targets that were hit by a dice roll */
+  private static Collection<Unit> findRolledTargets(
+      final List<Unit> availableTargets, final DiceRoll dice) {
+    final List<Die> rolls = dice.getRolls();
+    final List<Unit> hitTargets = new ArrayList<>();
+    for (int i = 0; i < rolls.size(); i++) {
+      if (rolls.get(i).getType() == DieType.HIT) {
+        hitTargets.add(availableTargets.get(i));
```

Comment:
It looks like there is an implicit assumption that avaialbleTargets will have the same size as 'dice.getRolls'. This assumption is not necessarily clear here, a `Precondition.checkArgument(availableTarget.size() == dice.getRolls().size())` at the start of the method would help document this. A reader would do a double take here wondering if there could be an indexing problem because `i` is tracking against `rolls`, not `availableTargets`.